pkg/ebpf/bytecode/runtime/runtime_compilation_helpers.go

```diff
@@ -44,166 +47,130 @@ var defaultFlags = []string{
+	baseName := strings.TrimSuffix(filename, filepath.Ext(filename))
 
+	unameHash, err := getUnameHash()
 	if err != nil {
-		return nil, err
+		return "", err
 	}
 
-	if _, err := os.Stat(outputFile); err != nil {
-		if !os.IsNotExist(err) {
-			rc.telemetry.compilationResult = outputFileErr
-			return nil, fmt.Errorf("error stat-ing output file %s: %w", outputFile, err)
-		}
-		dirs, res, err := kernel.GetKernelHeaders(config.EnableKernelHeaderDownload, config.KernelHeadersDirs, config.KernelHeadersDownloadDir, config.AptConfigDir, config.YumReposDir, config.ZypperReposDir)
-		rc.telemetry.headerFetchResult = res
-		if err != nil {
-			rc.telemetry.compilationResult = headerFetchErr
-			return nil, fmt.Errorf("unable to find kernel headers: %w", err)
-		}
-		if err := compiler.CompileToObjectFile(inputReader, outputFile, flags, dirs); err != nil {
-			rc.telemetry.compilationResult = compilationErr
-			return nil, fmt.Errorf("failed to compile runtime version of %s: %s", inputFileName, err)
-		}
-		rc.telemetry.compilationResult = compilationSuccess
-		log.Infof("successfully compiled runtime version of %s", inputFileName)
-	} else {
-		rc.telemetry.compilationResult = compiledOutputFound
+	outputFile := filepath.Join(outputDir, fmt.Sprintf("%s-%s-%s-%s.o", baseName, unameHash, inputHash, flagHash))
+	return outputFile, nil
+}
+
+func openCompiledOutput(filename, outputFile string) (CompiledOutput, error) {
+	if _, err := os.Stat(outputFile); errors.Is(err, os.ErrNotExist) {
+		return nil, fmt.Errorf("%s has not been compiled", filename)
 	}
 
-	err = bytecode.VerifyAssetPermissions(outputFile)
+	err := bytecode.VerifyAssetPermissions(outputFile)
 	if err != nil {
-		rc.telemetry.compilationResult = outputFileErr
 		return nil, err
 	}
 
 	out, err := os.Open(outputFile)
 	if err != nil {
-		rc.telemetry.compilationResult = resultReadErr
+		return nil, fmt.Errorf("failed to read compiled output: %w", err)
 	}
-	return out, err
+
+	return out, nil
 }
 
-// Sha256hex returns the hex string of the sha256 of the provided buffer
-func Sha256hex(buf []byte) (string, error) {
-	hasher := sha256.New()
-	if _, err := hasher.Write(buf); err != nil {
-		return "", err
+// getUnameHash returns a sha256 hash of the uname release and version
+func getUnameHash() (string, error) {
+	// we use the raw uname instead of the kernel version, because some kernel versions
+	// can be clamped to 255 thus causing collisions
+	var uname unix.Utsname
+	if err := unix.Uname(&uname); err != nil {
+		return "", fmt.Errorf("unable to get kernel version: %w", err)
 	}
-	cCodeHash := hasher.Sum(nil)
-	return hex.EncodeToString(cCodeHash), nil
-}
 
-// UnameHash returns a sha256 hash of the uname release and version
-func UnameHash(uname *unix.Utsname) (string, error) {
 	var rv string
 	rv += unix.ByteSliceToString(uname.Release[:])
 	rv += unix.ByteSliceToString(uname.Version[:])
-	return Sha256hex([]byte(rv))
+	return sha256hex([]byte(rv))
```

Comment:
Not sure why `string` is needed here
```go
sha256hex(append(uname.Release, uname.Version...))
```
We could also have 
```go
func sha256hex(bufs ...[]byte)
```
... and then we don't have to use `append`.